WalletWasabi/Crypto/StringCipher.cs

```diff
@@ -101,10 +101,15 @@ public static string Decrypt(string cipherText, string passPhrase)
 			memoryStream.Seek(-cipherLength, SeekOrigin.End);
 			using var cryptoStream = new CryptoStream(memoryStream, decryptor, CryptoStreamMode.Read);
 			var plainTextBytes = new byte[cipherLength];
-			var decryptedByteCount = cryptoStream.Read(plainTextBytes, 0, plainTextBytes.Length);
+			var read = 0;
```

Comment:
L110 and L111: Those `.Close()` method calls are not necessary because we use `using` to dispose those streams. The order of L110 and L111 is incorrect anyway, they should be called in opposite order I guess.

Other than that, I would expect an exception when we do not read as many bytes as expected. So something like this:

```cs
	int read = 0;
	int totalRead = 0;

	while ((read = cryptoStream.Read(plainTextBytes, totalRead, cipherLength - totalRead)) > 0)
	{
		totalRead += read;
	}

	if (totalRead != cipherLength)
	{
		throw new CryptographicException("Failed to read all plain text bytes.");
	}

	return Encoding.UTF8.GetString(plainTextBytes, 0, totalRead);
```

Reply:
Hm, but then it seems easier not to do that at all and just do:

```cs
public static string Decrypt(string cipherText, string passPhrase)
{
	byte[] cipherTextBytesWithSaltAndIv = Convert.FromBase64String(cipherText);
	byte[] key;
	byte[] iv;

	using MemoryStream memoryStream = new(cipherTextBytesWithSaltAndIv);
	int cipherStartIndex;
	int cipherLength;

	using (var reader = new BinaryReader(memoryStream, Encoding.UTF8, true))
	{
		byte[] salt = reader.ReadBytes(KeySize / 8);
		iv = reader.ReadBytes(KeySize / 8);
		byte[] authenticationCode = reader.ReadBytes(32);
		cipherStartIndex = (int)memoryStream.Position;
		cipherLength = (int)(memoryStream.Length - memoryStream.Position);
		byte[] cipher = reader.ReadBytes(cipherLength);

		using (Rfc2898DeriveBytes password = new(passPhrase, salt, DerivationIterations))
		{
			key = password.GetBytes(KeySize / 8);
		}

		using HMACSHA256 hmac = new(key);
		byte[] calculatedAuthenticationCode = hmac.ComputeHash(iv.Concat(cipher).ToArray());

		for (var i = 0; i < calculatedAuthenticationCode.Length; i++)
		{
			if (calculatedAuthenticationCode[i] != authenticationCode[i])
			{
				throw new CryptographicException("Message Authentication failed. Message has been modified or wrong password");
			}
		}
	}

	using Aes aes = CreateAES();
	using ICryptoTransform decryptor = aes.CreateDecryptor(key, iv);
	byte[] plainTextBytes = decryptor.TransformFinalBlock(cipherTextBytesWithSaltAndIv, cipherStartIndex, cipherLength);

	return Encoding.UTF8.GetString(plainTextBytes);
}
```